### ISSUE TITLE
Operate on WebAssembly?

### ISSUE THREAD

**Issue by fitzgen** [google/souper#323]
I don't see any references to it anymore, but my understanding was that Souper supported not just LLVM IR but also MSVC's IR. With that in mind, I think it would be exciting if Souper also supported WebAssembly as an IR (potentially hinting to [binaryen](https://github.com/WebAssembly/binaryen) what optimizations it is missing).

Brief intro to WebAssembly for the uninitiated (from http://webassembly.org/ ):

> WebAssembly (abbreviated Wasm) is a binary instruction format for a stack-based virtual machine. Wasm is designed as a portable target for compilation of high-level languages like C/C++/Rust, enabling deployment on the web for client and server applications.

[WebAssembly also has an extensive specification](https://webassembly.github.io/spec/core/index.html), which hopefully makes supporting it easier than it would otherwise be.

I'm not proposing anything concrete (yet?), just trying to gauge interest and get feedback.

Thoughts?

cc @kripken @sunfishcode

**Comment by regehr** [comment]
I'll need to think about this, but I'll just note right now that Souper doesn't support MSVC, but rather MSVC IR and Souper IR are similar enough that Gratian at Microsoft wrote a quick MSVC->Souper translator and then manually implemented a bunch of optimizations that it discovered. Also, we recently learned that the Mono compiler people used Souper similarly! But I don't have much detail about that yet. I assume they used their SSA IR but I don't know for sure.

As far as non-SSA IRs go, a while ago I started on a CompCert->Souper translator but got stalled for lack of time. I don't have a good sense for how easy it would be to do a WebAssembly->Souper translator.

**Comment by regehr** [comment]
Ok, I thought about this a bit more.

The first issue is getting Souper to understand WebAssembly, this is less trivial than other cases due to the stack machine. On the other hand, a very nice thing about conversion to Souper is that you can stop whenever you want, for example because you run into something tricky. I've never written code to symbolically execute stack machine code to make it into register code, but conceptually this doesn't sound that hard (again, as long as we can bail out when things get tricky).

As far as I know, after un-stack-conversion (is there a word for this?) the underlying operations should map nicely between WebAssembly and Souper, and Souper's synthesizer should find optimizations perfectly normally.

The tricky part, then, is translating the synthesized results back into WebAssembly optimizations. I suppose this would be 100% manual at first, and then there are probably things that make sense to automate.

So anyhow my initial impression is that we want an unstackifier that makes Souper IR out of webasm, and then we can proceed from there. I think this makes more sense than trying to shoehorn a stack machine into the Souper implementation.

**Comment by fitzgen** [comment]
Thanks for the thoughtful reply :)

One benefit I see of destackifying wasm into Souper IR is that experimentation developing that tool can happen out-of-tree and can move fast and loose.

Souper's ability to bail out on tricky constructs is nice, too. This enables incremental development, where we could get an MVP working and then start ratcheting up the percent wasm that is understood by the translator.

On the other hand, I also have a concern with this approach. Code size is super important for wasm because it is delivered over the network, and therefore potential size optimizations are particularly enticing. I fear that translating wasm into Souper IR won't help us discover "silly" wasm-specific size optimizations, like https://kripken.github.io/blog/binaryen/2018/04/13/a-silly-optimization.html

Maybe this fear is unfounded, and size optimizations aren't as specific to the instruction set and its encoding as I'm assuming? Or maybe translation into Souper IR just isn't the right approach for wasm size optimization, and further development of a wasm-specific super optimizer is a more promising avenue of investigation?

**Comment by regehr** [comment]
I have the same worry about size! Best we can do is give it a try, and maybe it'll turn out that re-stackifying the Souper output gets back the benefits?
Synthesis is a hard enough problem that I think it's worth exploring with Souper before doing more one-offs.
Random thought: I wonder if perhaps the webassembly typechecker/bytecode verifier/whatever contains the simple symbolic execution engine that we want for unstackifying webasm?

**Comment by kripken** [comment]
I agree this is definitely worth trying! I could look into creating a Binaryen pass to convert Binaryen IR to Souper IR, if that would be helpful? (Binaryen IR is almost identical to wasm, can be converted to and from it, and preserves size and performance aspects.) It might be simplest for me to convert to the Souper text format maybe, to get started?

About the size concern, yeah, that seems like it could be an issue. For example if the wasm => Souper IR conversion is inefficient, then Souper's proposed optimizations might be to remove those inefficiencies. We'll just have to make sure it's a good conversion :)

**Comment by regehr** [comment]
Alon, if you could hack something up that generates Souper text, this would be great! I'm happy to read code, play with crappy prototypes, etc. Also I'm happy to do the Souper runs (synthesizer can be finicky).

One thing to keep in mind is that one of Souper's most powerful features is its path conditions-- so if there's a way to generate those easily as part of the conversion, this will be especially helpful in generating optimizations (though applying the optimizations becomes less straightforward).

**Comment by kripken** [comment]
Great! I'll see what I can hack up.

Is the [Souper paper](https://arxiv.org/abs/1711.04422) the best place for docs on the IR text format?

About path conditions, naively it seems like that should be easy since we know the conditions of conditional branches etc. But if we need to do stuff like forward conditions through multiple conditional branches etc. (if the condition is still true in them) then I'm not sure offhand.

**Comment by regehr** [comment]
Argh we don't really have docs for Souper IR -- but it is really simple. I guess maybe ask me lots of questions and look at the test cases?
```
find souper/test -name '*.opt'
```

**Comment by regehr** [comment]
Re. path conditions, extracting the condition from only the most recent branch is probably the way to go.

**Comment by chenyang78** [comment]
Sorry for chiming in on this :) Beside *.opt as John says, I think souper/InstRef.rst is a good place in terms of Souper IR. Otherwise, you may have to end up scanning the relevant code. 

In terms of path condition (PC in Souper's terminology), I think knowing the conditions of conditional branches should be enough for the first version. You can skip the BlockPC part at the moment.

**Comment by regehr** [comment]
Ah, regarding docs/InstRef.rst, let me take a look and make sure it's up to date, has not been touched for a quite a while.

**Comment by kripken** [comment]
Thanks @regehr and @chenyang78, those look like what I need to get started!

**Comment by kripken** [comment]
Ok, I wrote up something that seems like it's starting to have the right general shape for simple things. Feedback welcome! Code is in a `souper` branch in binaryen, the new pass is [here](https://github.com/WebAssembly/binaryen/blob/souper/src/passes/Souperify.cpp).

To try it, you can do something like
```
wasm-opt input.wasm --flatten --simplify-locals-nonesting --souperify
```
(`flatten` flattens the IR; `simplify-locals-nonesting` simplifies it while keeping it flat; `souperify` runs the actual conversion). This will print out all the LHSes it can handle. Example [input](https://github.com/WebAssembly/binaryen/blob/souper/test/passes/flatten_simplify-locals-nonesting_souperify.wast) and [output](https://github.com/WebAssembly/binaryen/blob/souper/test/passes/flatten_simplify-locals-nonesting_souperify.txt) are in the test suite. For example, something conceptually like
```
int func(int x) {
  if (x & 1)
    x += 1;
  else
    x += 2;
  return x & 1;
}
```
can be written in wasm like this
```
    (if
      (i32.and
        (get_local $x)
        (i32.const 1)
      )
      (set_local $x
        (i32.add
          (get_local $x)
          (i32.const 1)
        )
      )
      (set_local $x
        (i32.add
          (get_local $x)
          (i32.const 2)
        )
      )
    )
    (return
      (i32.and
        (get_local $x)
        (i32.const 1)
      )
    )
)
```
and `--souperify` will emit 5 LHSes for it, the interesting one is for the returned value, which is
```
%0 = block 2
%1 = var
%2 = and %1, 1:i32
%3 = blockpc %0 0 %2 1:i32
%4 = blockpc %0 1 %2 0:i32
%5 = add %1, 1:i32
%6 = add %1, 2:i32
%7 = phi %0, %5, %6
%8 = and %7, 1:i32
infer %8
```
which *looks* like it might be valid Souper IR. I didn't try to load it yet though... :)

Still a lot of TODOs (IR not yet converted) and FIXMEs (no `i1` conversions yet). It just stops a particular LHS when it see stuff it can't handle. For control flow so far only `if`s are handled. `blockpc`s for them are implemented too (see example above), but no regular `pc`s yet. I suspect `if`s are enough control flow to get interesting results.

Some thoughts as I was doing this:
 * One thing I would expect the Souper approach to help with is "backend" inefficiencies. wasm is often code that has been run through the LLVM optimizer (nothing new there for Souper), but then it goes through an LLVM backend, which maybe introduces extra work there.
 * wasm **not** run through the LLVM optimizer is interesting too, e.g. generated from Go or AssemblyScript, and maybe optimized by Binaryen. Souper on such wasm could find stuff the Go, AssemblyScript, and Binaryen optimizers could do better. I'm guessing there is quite a lot there!
 * We can't expect this to handle very low-level wasm things like [-(-64) being smaller than 64](https://kripken.github.io/blog/binaryen/2018/04/13/a-silly-optimization.html), or things with `tee`s, `block` and `if` values, etc. All those details get lost in the translation to Souper IR. Overall I expect we can benefit both from the Souper approach and from [superoptimizing on wasm directly](https://github.com/WebAssembly/binaryen/pull/900) (I recently got that building again for more tests), they are probably complementary.

**Comment by regehr** [comment]
Awesome. Is there a particular C/C++ code base that is particularly representative that I should start with? Otherwise I'll just build a few random things.

So far I've been creating wasm by passing --target=wasm32 to clang, is that the right approach?

Indeed, Go's equivalent of InstCombine is quite weak so Souper will find a lot of stuff.

There's an interesting piece of this puzzle that I don't think has been attacked seriously yet, that would not only be useful to solve but also would make a good submission to a conference like PLDI. This is taking patterns from Souper, perhaps abstracting them so they're more broadly applicable, and then making a very fast matcher for them. Just mentioning this since it's something I'm very interested in working on.

**Comment by kripken** [comment]
> Awesome. Is there a particular C/C++ code base that is particularly representative of code you're interested in optimizing, that I should start with? Otherwise I'll just build a few random things.

Hmm, some good samples are here: https://github.com/kripken/embenchen/tree/master/asm_v_wasm (that's output from the emscripten benchmark suite). In particular the larger ones like bullet, box2d, and lua, those are important real-world codebases I think.

> So far I've been creating wasm by passing --target=wasm32 to clang, is that the right approach?

In general yes, however it's worth running the binaryen optimizer on it too (`wasm-opt -Os input.wasm -o output.wasm`) as it will shrink it by 10% or so. Probably not the same optimizations as Souper would look for, but while shrinking it makes the code cleaner which is more likely to work in the `--souperify` pass.

**Comment by regehr** [comment]
Thanks Alon. I should have a bit of time this week to work on this!

**Comment by kripken** [comment]
`pc` implemention added to that branch. Output looks like it might be correct, but I'm not sure :)

**Comment by kripken** [comment]
@regehr thanks! I fixed those 3 things on the branch.

Output looks better now, but probably still a lot left to do to get it right. Also, to add operations aside from binaries (unaries, etc.)...

**Comment by regehr** [comment]
Ok, great! 
I'm now seeing two small issues.

First, while most LHSs have the proper var declarations there are also some degenerate LHSs that are just this, which makes Souper's parser barf:
```
; start LHS
infer %0
```

Then second there's a common pattern of comparing the results of comparisons against zero using the wrong width as seen here:
```
; start LHS
%0:i32 = var
%1 = slt -1000999:i32, %0
%2 = sle 0:i32, %0
%3 = eq %2, 0:i32
pc %3 1:i1
infer %1
```
Since Souper doesn't have any implicit sext/zext, it should be:
```
%3 = eq %2, 0:i1
```

But leaving aside these issues I'm already seeing some perhaps-useful stuff coming out of the synthesizer!

**Comment by regehr** [comment]
As a random example, here's an optimization that Souper would like to perform.
LHS:
```
%0:i32 = var
%1 = sle 0:i32, %0
%2:i32 = zext %1
%3:i32 = var
%4 = sle 0:i32, %3
%5:i32 = zext %4
%6 = or %2, %5
%7 = eq %6, 0:i32
infer %7
```
RHS:
```
%8:i32 = and %0, %3
%9:i1 = slt %8, 0:i32
result %9
```

**Comment by kripken** [comment]
Exciting to see results already! And nice optimization it found there.

Those two issues should be fixed on the branch. Also added unary ops (cttz etc.).

**Comment by regehr** [comment]
Ok!
Next two issues should also be easy:
- please use lshr/ashr instead of sshr/ushr
- blockpc doesn't get a label in Souper IR, so instead of:
```
%3 = blockpc %0 0 %2 1:i1
```
we want:
```
blockpc %0 0 %2 1:i1
```

**Comment by regehr** [comment]
And here are a few more results:
https://gist.github.com/regehr/a428f164ae60ff783e5872d5efc5416e

To get these, I took wasm_lua_scimark.c.wasm and ran it through binaryen -O3, then ran the output through Souper in "constant synthesis" mode, where the only thing it does is tries to prove at the LHS is a constant.

I haven't gone and tracked these back to the wasm, so some of them could be artifacts of the translation, but it looks like at least some of them represent real opportunities for shaving code size.

To facilitate backtracking, would it make sense for --souperify to also print the wasm in a comment?

**Comment by regehr** [comment]
You'll notice that sometimes the LHS contains far more context than is required to justify the optimization, making it harder to understand what's going on. I have a reducer for LHSs that I can run in the future.

**Comment by kripken** [comment]
Thanks! Fixed those two minor issues on the branch.

About backtracking, I need to think about that - we have a DebugInfo mechanism that might work somehow, but it's not trivial to connect it. Looks like we definitely need a way to do that for debugging, though, as several of those results do look like my translation is wrong, for example the optimizer definitely knows that multiplying an integer by zero is zero, and can add two constants, etc.

**Comment by kripken** [comment]
To get something simple I added support for setting `BINARYEN_DEBUG_SOUPERIFY=1` in the env, then it prints each wasm expression before the souper one. However, it's not in a comment (no trivial way do that atm, I'll look into it) but worse it's printing the "flattened" IR, so it may not be useful enough for real debugging.

However, when trying to use this to debug those weird muls and adds, I think I might be misreading your gist:
```
; RHS inferred successfully
%0:i32 = add 0:i32, 1:i32
cand %0 1:i32

; RHS inferred successfully
%0:i32 = var
%1:i32 = mul 0:i32, %0
cand %1 0:i32
```
Does that mean I should be able to search in the output from processing that lua file (after `-O3`) and find the strings ` = add 0` or ` = mul 0` in the LHSes? Neither seems to show up. All the adds and muls have a non-constant left side (starting with `%`) as I'd expect, so maybe I'm confused? Or was this another file than https://github.com/kripken/embenchen/blob/master/asm_v_wasm/wasm_lua_scimark.c.wasm ?

**Comment by regehr** [comment]
I'm getting an abort on the latest binaren/souperify:
```
~/embenchen/asm_v_wasm$ ~/binaryen/build/bin/wasm-opt -O3 wasm_lua_scimark.c.wasm -o souper-input.wasm
~/embenchen/asm_v_wasm$ ~/binaryen/build/bin/wasm-opt souper-input.wasm --flatten --simplify-locals-nonesting --souperify > lhs.opt
Aborted
```

**Comment by regehr** [comment]
You're not misreading the gist, Souper legit thinks that there's a mul x, 0 in there. I'll get more details.

**Comment by kripken** [comment]
Sorry about the abort - stability has decreased as I added more features (`select` etc.) - it's converting a lot more code now so it's hitting more corner cases. I pushed some bugfixes now, which fix `wasm_lua_scimark.c.wasm` and a bunch of other large codebases.

**Comment by regehr** [comment]
OK! Still working with a wasm_lua_scimark.c.wasm that has been passed through binaryen -03, I now see only one species of souper error, a width mismatch on code like below. Since the result of a Souper comparison is i1, %4 should be extended to i32 before passing it into the phi node.

```
; start LHS
%0 = block 2
%1:i32 = var
; (i32.or(get_local $15)(i32.const 32))
%2 = or %1, 32:i32
; (i32.add(get_local $16)(i32.const -97))
%3 = add %2, -97:i32
; (i32.lt_u(get_local $17)(i32.const 26))
%4 = ult %3, 26:i32
%5 = phi %0, %1, %4
infer %5
```

**Comment by regehr** [comment]
Interesting-- most of the constant optimizations have now gone away, hopefully the missing ones were just the bogus ones!
See the remaining optimizations here:
https://gist.github.com/regehr/cd0d7a4c1f99226f4a78571bb72df0f4
Synthesizing instructions is slower, I'll start that going now.

**Comment by kripken** [comment]
Ok, fixed that phi/i1 bug.

Looking at those latest optimizations,
```
%1 = phi %0, 7:i32, 7:i32
```
That seems like it might be a bug in my phi generation, I'll look into it.

**Comment by kripken** [comment]
Souper found something important there! :) It wasn't a bug in my phi generation, it was a real missing optimization. Fixed in https://github.com/WebAssembly/binaryen/pull/1512 (more details there) and also included in the `souper` branch.

After that, I don't see any more trivial phis with identical constant inputs (although, it's possible more exist due to a different cause somehow).

**Comment by regehr** [comment]
Nice! I'll followup with results from synthesis, but probably after a delay since I have a trip later this week.

**Comment by kripken** [comment]
@regehr when you get a chance, a question: I noticed I was halting LHS generation early on anything bad, like a function call, so I switched that to emit a `var` instead for such values. Now it's generating really huge LHSes for really huge functions, e.g. in `qt.wasm` some are thousands of lines long (!) Is there a practical size limit for LHSes that I should stop at?

**Comment by regehr** [comment]
OK, the question "how big to make the LHS" gets into some thorny details.

Answer #1 is that you want the LHS as large as possible since this gives the solver the most information. In the simple cases -- constant synthesis and "nop synthesis" where the output is already available somewhere in the input graph -- this is always the right answer, unless it overloads the solver to the point where timeouts/memoryouts become an issue.

Answer #2 is not as fun because of how Souper synthesis currently works. Synthesis has to make an entire RHS from scratch, it cannot reuse parts of the LHS. SInce synthesis scales to maybe 5-7 instructions, a larger LHS is unlikely to have any equivalent RHS that is within practical reach. We've known about this issue for a while and unfortunately solving it requires some deep changes to Souper. We (or more to the point @rsas) have implemented many of the necessary pieces, but they're not all there yet. In the meantime, instruction synthesis may succeed if you extract only up to some smaller depth, for example graph height in the range 2-4. 

So, the long-term answer is: put as much as you can into Souper. The short term answer is that we'd like a configurable depth-bound in the converter from X to Souper. We have such a thing for LLVM IR, though it only lives in a branch. Anyhow this should be an easy feature to add -- if you do that, I'll play around with various depth limits.

Sorry to not have a nicer answer!

**Comment by regehr** [comment]
Overall, of course, turning a Souper-stopper into a var, as opposed to bailing out entirely, is absolutely the correct answer.

**Comment by regehr** [comment]
Randomly,  if you happen to have linkage info, there's no reason not to peek into called functions and get more instructions from there. It's one of many things I've wanted to try but haven't had time for. Perhaps it doesn't matter much once we've done a good job inlining, though. Certainly a bit of restraint would be called for, or else LHSs would become really big.

**Comment by regehr** [comment]
Ok, keeping on going with that same .wasm file, here are the constants and nops:
https://gist.github.com/regehr/c86af851d60a75e04ce19e270bcfc554
I think some of these are actionable, but definitely not all.

At some point we should talk about undefined behavior. Souper is going by LLVM's rules, which are a bit looser than webassembly's, particularly for shift instructions. This will eventually give us some spurious optimizations. We don't have any kind of pluggability for instruction semantics in Souper, so will have to figure out how to best deal with this issue. Possibly just a one-off hack. Anyhow I'm going to wait until the problem shows itself before thinking more seriously.

**Comment by regehr** [comment]
One other thing: I'd like to compare the optimizations coming out of the .wasm files with the optimizations coming out of the corresponding .bc files. I don't see those in the embenchen repo, can you let me know where I might find them?

**Comment by regehr** [comment]
Here are a few synthesis results:
https://gist.github.com/regehr/b7b990d7c49cec51e14c646660cfa240

And a few more things we should talk about:
- Some of these optimizations are probably legit, others will likely be spoiled by uses of intermediate values that aren't apparent in the Souper code that you generated. This is an issue that we have in LLVM as well. The solution we're working on is to make these "external uses" visible in Souper IR at which point the synthesizer will take this into account. (This is the simple solution; it has the unfortunate property of allowing collections of interdependent optimizations to block each other. We know how to fix this as well but it'll be a bit more work.)
- Souper has a cost model that it uses to determine whether a candidate RHS is better than its LHS. This cost model is somewhat guesswork since costs are not totally clear at the IR level. Moreover, the cost model that we've come up with is probably not right for WebAssembly. This is easy to tweak. If you want to look at our current (extremely simple) cost model, it is here: https://github.com/google/souper/blob/master/lib/Inst/Inst.cpp#L640

**Comment by regehr** [comment]
One more thing: as I mentioned earlier, I have a minimizer for LHSs that pares them down to the smallest code that justifies the given optimization. Let me know if you'd like me to run this before giving you results. It will have the unfortunate effect of ruining the backtracking information that you put into the generated Souper, so perhaps it would make sense to present both minimized and non-minimized versions.

Making Souper optimizations understandable to compiler developers is something I've put a lot of thought into, but I'd be happy to hear more thoughts about this. In the long run it is a non-issue since Souper will be turned into table-driven optimizers that humans won't have to look at :)

**Comment by kripken** [comment]
Makes sense about the depth parameter, thanks, I'll look into adding that.

Very interesting point about inlining, it seems not too hard to look into called functions (up to a depth limit or such), if we keep around the DataFlow IR for them. Worth looking into, I agree.

I updated all the builds in https://github.com/kripken/embenchen/tree/master/asm_v_wasm , and now they also have the bitcode files there. There are also separate wasm versions for asm2wasm and the LLVM wasm backend - aside from LLVM version differences, those should differ on only the backend conversion of LLVM IR to wasm, so any differences in the optimizations Souper finds would be interesting.

> Some of these optimizations are probably legit, others will likely be spoiled by uses of intermediate values that aren't apparent in the Souper code that you generated.

I don't understand this, sorry?

>  In the long run it is a non-issue since Souper will be turned into table-driven optimizers that humans won't have to look at :)

That's the future I hope for :) I'd like to import large amounts of superoptimizer-generated optimizations in an automatic matter. I guess this ties into the fast matcher idea you mentioned before - I'm also very interested in that, interesting to think about.

Meanwhile I looked into more of the suspiciously trivial stuff we were emitting on lua, and after the two Binaryen PRs all that was left was a serious bug in phi generation, that turned out to need some large refactoring of the code. But now that that's done (pushed to the main branch), I don't see us emitting anything suspiciously trival on lua, and I think our phi handling is now correct.

Looking at your latest results, one optimization is 24 + 36 = 60, which also looks like a trivial thing that indicates either a bug or a missing optimization, I'll investigate...

**Comment by kripken** [comment]
Hmm, I don't see that 24 + 36 = 60 thing on lua, so perhaps it was only there because of the phi issue that was just fixed?

**Comment by regehr** [comment]
I'll rerun soon (but perhaps not until early next week, I'm leaving town early tomorrow).

Thanks for adding bitcode! It'll be really interesting comparing results across bitcode and wasm. I assume I can turn off inlining when I optimize wasm, to make it easier to map between the two?

Re. table-driven optimizers let's talk sometime. I think this is an important step forward in how we build these things. I'm happy to come out and spend half a day with y'all sometime.

Regarding external uses, see the bad diagram below. It shows a LHS with two add instructions that Souper would like to turn into a single add, for a win of 1 instruction. However, x+7 has an "outside use" that Souper doesn't capture, so if we implement the optimization we end up with a win of zero. We have some work in progress to capture these in Souper IR, in order to use the presence of outside uses to constrain the synthesizer. The analogous constraints in LLVM IR are the ubiquitous hasOneUse() checks that you see in InstCombine. Hope that made more sense this time.

![img_20180426_153121](https://user-images.githubusercontent.com/732536/39335242-71e3fc6a-496f-11e8-9e13-279f0ff83bd4.jpg)

**Comment by kripken** [comment]
> I assume I can turn off inlining when I optimize wasm, to make it easier to map between the two?

Those bitcode files are after LLVM inlining, but Binaryen may inline. To make it easy to map to the IR, I added a subdir `with_names` over there that has builds with function names in the binaries (which then show up in the `--souperify` output).

> Re. table-driven optimizers let's talk sometime. I think this is an important step forward in how we build these things. I'm happy to come out and spend half a day with y'all sometime.

Sounds good!

> Regarding external uses, see the bad diagram below.

I see now, thanks. Tricky, and, there are probably more such issues with wasm than with LLVM IR, as there are all kinds of structural effects from the surrounding code.

**Comment by kripken** [comment]
Oh, and I added a depth limit env var, `BINARYEN_SOUPERIFY_DEPTH_LIMIT` (default 10).

**Comment by kripken** [comment]
Added loop support: when there should be a `phi` at the top of a loop, a `var` is placed instead for correctness.

Before this change we just skipped functions with loops, so it is emitting a *lot* more LHSes now. I added a limit on the total instructions in an LHS, `BINARYEN_SOUPERIFY_TOTAL_LIMIT`, default to 30.

**Comment by regehr** [comment]
Nice! For now I will not change the default settings of BINARYEN_SOUPERIFY_DEPTH_LIMIT or BINARYEN_SOUPERIFY_TOTAL_LIMIT, and I'll keep working with lua_scimark, and I'm using the wasmbackend version in the with_names subdirectory. Hope that all sounds good and let me know what I can do to help make the results below more actionable.

These are just constants and nops, I'll follow up with a version where the synthesizer makes instructions:
https://gist.github.com/regehr/74bc92ad72f663bba3b23362ed7adc94

Those pairs of add/sub of multiples of 8 are interesting! I wonder what's going on with that? They are definitely not there in the original bitcode.

For comparison, here are the results from running Souper on the corresponding bitcode file:
https://gist.github.com/regehr/43133fbaec3e6454aa64225c522f5b95

**Comment by kripken** [comment]
Yeah, those add/sub pairs look very suspicious :) I wonder if it's not from lowering of GEPs that somehow isn't getting optimized. Oddly I can't seem to see them when I run something like
```
BINARYEN_DEBUG_SOUPERIFY=1 bin/wasm-opt ../bench/asm_v_wasm/with_names/asm2wasm_lua_scimark.c.wasm -O3 --flatten --simplify-locals-nonesting --souperify
```

Btw, which `lua_scimark` is that? If it's https://github.com/kripken/embenchen/blob/master/asm_v_wasm/with_names/asm2wasm_lua_scimark.c.wasm (or another from that directory) then it should have names, and would print stuff like
```
; start LHS (in $_match)
```
instead of 
```
; start LHS (in $56)
```

**Comment by regehr** [comment]
I used with_names/wasmbackend_lua_scimark.c.wasm, should I start again with the asm2wasm one?

**Comment by regehr** [comment]
Is it OK for me to invoke wasm-opt just once, as you did, like this?
```
-O3 --flatten --simplify-locals-nonesting --souperify
```
It isn't always clear to me how flags compose and I had been running it twice to make sure that the -O3 runs before the souperify.

**Comment by kripken** [comment]
I see what's going on, yeah - if you ran two commands, then binaryen by default won't save function names (you need `-g` to preserve them). And yeah, it's fine to compose those things, `-O3`, `--souperify` etc. are all just passes (or lists of passes), run in order, so a single command like my example is fine too.

**Comment by kripken** [comment]
And yeah, now I see those trivial add/sub pairs - they are in the wasmbackend versions, not the asm2wasm ones. E.g.
```
; start LHS (in $luaL_error)
%0:i32 = var
; (i32.sub(get_local $5)(i32.const 128))
%1 = sub %0, 128:i32
; (i32.add(get_local $3)(i32.const 128))
%2 = add %1, 128:i32
infer %2
```

**Comment by kripken** [comment]
And here's the relevant wast:
```
 (func $luaL_error (; 41 ;) (type $3) (param $var$0 i32) (param $var$1 i32) (param $var$2 i32) (result i32)
  (local $var$3 i32)
  (local $var$4 i32)
  (i32.store
   (i32.const 1024)
   (tee_local $var$3
    (i32.sub
     (i32.load
      (i32.const 1024)
     )
     (i32.const 128)
    )
   )
  )
  [..]
  (i32.store
   (i32.const 1024)
   (i32.add
    (get_local $var$3)
    (i32.const 128)
   )
  )
  (get_local $var$0)
 )
```
Looks like stack prologue/epilogue stuff - things added by the wasm backend, after LLVM IR.

I think that could be improved, but I'm not sure. Opened https://bugs.llvm.org/show_bug.cgi?id=37299

**Comment by regehr** [comment]
Interesting!
I'll do the next synthesis run with instructions on the asm2wasm version.
Unfortunately it may not finish before I leave town again, and my current fastest machine that I use for synthesis happens to be at home, so it may be the weekend before I post results.
Let me know if it would be useful to do a constant/nop synthesis run on the asm2wasm version, I can turn that around quickly.
I have money for faster machines at work but haven't gotten around to ordering anything, argh.

**Comment by regehr** [comment]
Here's an example where we need to ask what the cost model should be. Souper wants to eliminate an instruction, but is that an appropriate optimization here?
```
; start LHS (in $_lua_settop)
%0:i32 = var
; (i32.le_s(get_local $1)(i32.const -1))
%1 = sle %0, -1:i32
%2:i32 = zext %1
infer %2

; RHS inferred successfully
%3:i32 = lshr %0, 31:i32
result %3
```

**Comment by regehr** [comment]
Some constants and nops:
https://gist.github.com/regehr/649c148d8c154b2550062d865b30a244
From:
```
BINARYEN_DEBUG_SOUPERIFY=1 ~/binaryen/build/bin/wasm-opt asm2wasm_lua_scimark.c.wasm -O3 --flatten --simplify-locals-nonesting --souperify
```

**Comment by regehr** [comment]
Does it seem weird that Souper gets 16k LHSs out of asm2wasm_lua_scimark.c.js.bc, but --souperify gets 200k LHSs out of asm2wasm_lua_scimark.c.wasm?

**Comment by regehr** [comment]
Well, in the ongoing instruction synthesis run, I'm getting quite a bit of this. Something I'd like to do is make a custom cost function for Souper that closely mirrors the actual code size costs in the webassembly  binary format.

I gather that some of the sexts/zexts in --souperify output are zero-cost in the sense that they were added only to make Souper happy. One thing we might consider is marking (in --souperify output) which instructions are freebies that have zero cost in the original wasm file. I could hack Souper to take this into account when it attempts to make a profitable optimization. This, plus some handling of external uses, might significantly reduce the noise level in synthesis results.

```
; start LHS (in $_lua_tolstring)
%0:i32 = var
; (i32.le_s(i32.const 0)(get_local $68))
%1 = sle 0:i32, %0
%2:i32 = zext %1
; (i32.eq(unreachable)(unreachable))
%3 = eq %2, 0:i32
infer %3

; RHS inferred successfully
%4:i32 = ashr %0, 31:i32
%5:i1 = trunc %4
result %5
```

**Comment by kripken** [comment]
Does "synthesis" mean stuff aside from const/nop?

In general results on the wasm backend are more important for the future (as we plan to switch to it from asm2wasm), however right now it is less optimized so more trivial stuff is possible there. But that's trivial stuff we need to fix ;)

Yeah, the cost model seems like it matters in that example. For code size, a `zext` of an `i1` is free in wasm. On the other hand, I'm not sure it's free in the compiled machine code? So depending on if we're optimizing for download size or runtime speed could matter.

**Comment by regehr** [comment]
Yeah, sorry, "synthesis" means making new instructions that weren't there before.

The issue of cost models is one that we've spent a lot of time thinking about. I think the issues are basically the same for LLVM IR and WebAssembly. Options include:
- right now we just eyeball it, tweaking the model when synthesis comes up with something it shouldn't have. this is only minimally workable. LLVM has these somewhat complicated and poorly documented ideas about canonicalization and they're hard to capture in a simple cost model.
- probably the next thing I'll try is to run a ton of small LLVM functions through opt and then build a cost model by observing LLVM's behavior. I'm not certain that this approach makes sense for WebAssembly since the optimizer isn't as thorough yet.
- I also want to run a ton of small functions through a backend, apply an assembly-level cost model, and then try to map that back to the LLVM level. this should work really well when optimizing for the size of the eventual object code; it isn't clear how well it'll work when optimizing for throughput, since predicting throughput of assembly code is difficult for modern processors. we can try this approach for webassembly assuming that you have an ahead-of-time webassembly backend?
Hope this all makes sense. Summary is that this is a problem we need to solve anyway, so happy to put some effort into it.

**Comment by kripken** [comment]
Interesting about the cost model. Yeah, sounds like a lot to do there.

Meanwhile I generated some LHSes that might be interesting, this is from a function in the entropy coder in the AV1 media codec, which is something they tell me is relevant for superoptimization:
https://gist.github.com/kripken/b5da8f8da6b3cb88de012aa24e3393ce Would be interesting if it finds any especially clever stuff there.

I've also been improving the LHS generation mechanism. I'm working on a fuzzer that will verify that we generate proper LHSes from wasm, by doing trivial operations on them that should change nothing and then running to see that the output doesn't change. Already found one bug.

**Comment by regehr** [comment]
Great! I'm traveling and won't do anything on this until the weekend unfortunately. Looking forward to seeing the new stuff. Instruction synthesis hadn't finished before I left home this morning argh.

**Comment by regehr** [comment]
Results from the AV1 LHSs linked above:
https://gist.github.com/regehr/be7fd81ee4940086fcbf1fa3dd521fe0
The last one (with bswap) might be cute for processors where bswap isn't microcoded.

**Comment by kripken** [comment]
Thanks!

Reading the lua synthesis results, one problem is that we tell Souper to infer the artificial instructions that we add just for legal reasons, like `zext`s and comparisons to zero for path conditions. I pushed an improvement to avoid that.

Another issue is that the cost of `zext` and `trunc` should be 0. Not inferring those instructions (last paragraph) will help but they can also appear in the middle of LHSes. But, maybe that will be rare enough it's not a problem.

A third issue is the presence of additional uses not seen in the LHS, as you brought up before. I looked into a couple manually, and in all of them that was why Binaryen didn't already perform the optimization. I wonder if this is more of an issue on wasm than on LLVM IR for some reason. In any case, I wonder if it wouldn't be more interesting to focus on single-use instructions, and I added a pass for that, `--souperify-single-use` (can use it just like `--souperify`).

Thinking about this, the single-use case might be interesting to focus on for an additional reason, that it should be much easier to create a fast matcher for such patterns - in particular, such a matcher could operate on Binaryen IR directly as opposed to first constructing the DataFlow IR. That should be far faster.

Reading the AV1 results, those are very interesting :) I'll pass those along.

**Comment by regehr** [comment]
Great, I'm glad we're at the point of worrying about these sorts of details.

The current Souper cost model is dead simple:
https://github.com/google/souper/blob/master/lib/Inst/Inst.cpp#L640

I'll make an alternate version that has cost of ext/trunc at 0.

Regarding external uses: empirically, the single-use restriction will be a serious one, though certainly we should pursue the low hanging fruit there. But soon I'd like to move to a more sophisticated model where you mark instructions with external uses in the Souper IR and then synthesis will take them into account. The syntax for this isn't worked out yet but we'll do this soon. There's room for even more improvement beyond that, but let's not worry about that right now.

**Comment by regehr** [comment]
Using the latest binaryen, I've just kicked off a synthesis run for these LHSs:
```
BINARYEN_DEBUG_SOUPERIFY=1 ~/binaryen/build/bin/wasm-opt wasmbackend_bullet.wasm -O3 --flatten --simplify-locals-nonesting --souperify-single-use
```

; start LHS (in $free)
%0:i32 = var
%1:i32 = var
; (i32.ne(get_local $36)(get_local $2))
%2 = ne %0, %0
infer %2

; RHS inferred successfully
result 0:i1

**Comment by regehr** [comment]
OK, feel free to generate Souper that looks like this:
```
%0:i32 = var (hasExternalUses)
%1:i32 = add 0:i32, %0
%2:i32 = add 0:i32, %1
%3:i32 = add 0:i32, %2 (hasExternalUses)
%4:i32 = add 0:i32, %3
```
We'll modify synthesis to treat externally used values as fixed, and not try to replace them.
Please keep in mind that:
- checking for a single use is not the right algorithm, since many values have multiple uses that are captured by the Souper LHS; we really want to only mark values that have uses that are not visible in the Souper code
- the root of a LHS can obviously have as many uses as it likes and all of them are, by definition, invisible to Souper, and these have no impact on synthesis or the subsequent optimization

**Comment by regehr** [comment]
Great! I'll kick off a new synthesis run tonight but then I'm leaving town tomorrow so it'll be the weekend again before I see the results (maybe I should poke holes in my home firewall but I don't really want to).

**Comment by regehr** [comment]
here are a few synthesis results from wasmbackend_bullet.wasm:
https://gist.github.com/regehr/2378313b8b4ed859948341e79c375eeb

**Comment by regehr** [comment]
Also here's souper's cache dump of the same information (which I find easier to read without the webasm noise): https://gist.github.com/regehr/152b5e8359509c52a6c36829506d9d05

**Comment by kripken** [comment]
Thanks!

Seeing some trivial stuff that shouldn't be there, I found that the single-use calculation was actually wrong - it ignored uses in things like stores and calls. Fixed, hopefully now it's correct.

**Comment by kripken** [comment]
`--souperify` now emits `(hasExternalUses)` when relevant. This required some redesign of the IR to properly handle a single wasm instruction that turns into multiple Souper instructions, etc.

(`--souperify-single-use` still exists, and should never emit that notation, which is useful for fuzzing)

I think that's all I had on my list of stuff to fix.

**Comment by regehr** [comment]
Thanks! I'll post some new synthesis results soon.

**Comment by shrin18** [comment]
is there any reliable way to convert .opt to .ll

**Comment by regehr** [comment]
@shrin18 please explain in more detail what you want to do

**Comment by shrin18** [comment]
Sorry @regehr .

1. __`*.wat`__`(WASM text format)` →  __`*.o`__`(WASM binary file)`
2. __`*.o`__`(WASM binary file)` → __`*.bc`__`(LLVM bitcode file)`
3. __`*.bc`__`(LLVM bitcode file)` → __`*.opt`__`(souper candidates)`
4. __`*.opt`__`(souper candidates)` → __`*.ll`__`(LLVM IR)`
5. __`*.ll`__`(LLVM IR)` → __`*.o`__`(WASM binary file)`

We are implementing this pipeline. We can get all the steps until the 3rd step however we are stuck with the 4th step. We tried to use the utils/souper2llvm script in Souper. However it does not give us the expected output even when we pass the successful candidates as input. 

Further we also tried to use 
```
$ /path/to/clang -Xclang -load -Xclang /path/to/libsouperPass.so \
                 -mllvm -z3-path=/usr/bin/z3 /path/to/file.c
```
but we think that it is not meant to give an output in .ll format. What are we expected to do in this case if we want output in the .ll format from the optimal candidates. 

We are in the progress of experimenting with lib/Pass.cpp currently.

**Comment by shrin18** [comment]
when we use `BINARYEN_DEBUG_SOUPERIFY=1 ~/binaryen/build/bin/wasm-opt wasmbackend_bullet.wasm -O3 --flatten --simplify-locals-nonesting --souperify-single-use` we get LHS  for eg `; start LHS (in $0)
%0:i32 = var
%1:i32 = var
%2 = shl %1, 4:i32
%3 = add %0, %2 (hasExternalUses)
%4:i32 = var
%5 = sub %3, %4
%6 = add %5, 15:i32
%7 = add %3, 16:i32
%8 = ult %4, %7
%9 = sle 0:i32, %1
pc %8 1:i1
pc %9 1:i1
infer %6`

How do we get the RHS candidates from this ?
